cli/listener/src/main.rs

```diff
@@ -0,0 +1,256 @@
+    slot: Slot,
+    /// Epoch when the data point was logged.
+    epoch: Epoch,
+    /// Pool identifier, e.g. for Solido would be "solido".
+    pool: String,
+    /// Token_a price.
+    token_a: u64,
+    /// Token_b price.
+    token_b: u64,
+}
+
+pub fn create_db(conn: &Connection) -> rusqlite::Result<()> {
+    // The timestamp is stored in ISO-8601 format.
+    conn.execute(
+        "CREATE TABLE IF NOT EXISTS exchange_rate (
+                id          INTEGER PRIMARY KEY,
+                timestamp   TEXT,
+                slot        INTEGER NOT NULL,
+                epoch       INTEGER NOT NULL,
+                pool        TEXT NOT NULL,
+                token_a     INTEGER NOT NULL,
+                token_b     INTEGER NOT NULL
+            )",
+        [],
+    )?;
+    Ok(())
+}
+
+pub struct IntervalPrices {
+    t0: chrono::DateTime<chrono::Utc>,
+    t1: chrono::DateTime<chrono::Utc>,
+    epoch0: Epoch,
+    epoch1: Epoch,
+    price0_lamports: Rational,
+    price1_lamports: Rational,
+}
+
+impl IntervalPrices {
+    pub fn duration_wall_time(&self) -> chrono::Duration {
+        self.t1 - self.t0
+    }
+    pub fn duration_epochs(&self) -> u64 {
+        self.epoch1 - self.epoch0
+    }
+    pub fn growth_factor(&self) -> std::result::Result<Rational, ArithmeticError> {
+        self.price1_lamports / self.price0_lamports
+    }
+    pub fn annual_growth_factor(&self) -> f64 {
+        let year = chrono::Duration::days(365);
+        self.growth_factor()
+            .expect("Overflow happened when calculating growth factor.")
+            .to_f64()
+            .powf(year.num_seconds() as f64 / self.duration_wall_time().num_seconds() as f64)
+    }
+    pub fn annual_percentage_rate(&self) -> f64 {
+        self.annual_growth_factor().mul_add(100.0, -100.0)
+    }
+}
+
+pub fn get_apy_for_period(
+    conn: &Connection,
+    opts: &Opts,
+    from_time: chrono::DateTime<chrono::Utc>,
+    to_time: chrono::DateTime<chrono::Utc>,
+) -> rusqlite::Result<Option<f64>> {
+    let row_map = |row: &Row| {
+        let timestamp_iso8601: String = row.get(1)?;
+        Ok(ExchangeRate {
+            id: row.get(0)?,
+            timestamp: timestamp_iso8601
+                .parse()
+                .expect("Invalid timestamp format."),
+            slot: row.get(2)?,
+            epoch: row.get(3)?,
+            pool: row.get(4)?,
+            token_a: row.get(5)?,
+            token_b: row.get(6)?,
+        })
+    };
+
+    // Get first logged minimal logged data based on timestamp that is greater than `from_time`.
+    let mut stmt_first = conn.prepare(
+        "SELECT *, MIN(timestamp) FROM exchange_rate WHERE pool = :pool AND timestamp > :t",
+    )?;
+    let mut row_iter = stmt_first.query_map([opts.pool.clone(), from_time.to_string()], row_map)?;
+    let first = row_iter.next();
+
+    // Get first logged maximal logged data based on timestamp that is smaller than `to_time`.
+    let mut stmt_last = conn.prepare(
```

Comment:
If we issue multiple reads, we should do so in a transaction, to make sure that we get a consistent view of the data. A nice way to enforce that is to have an RAII wrapper type for transactions, and accept that as a function argument, instead of `Connection`. (This also makes it compose better: suppose you wanted to read the APY over different time periods. If this function would start the transaction itself, you still couldn’t do atomic reads of multiple different time periods. But if we accept the transaction as an argument, then they can all happen within the same transaction. (SQL does allow nested transactions, but the syntax is different :hankey: Imagine that in Rust you had to open the first block in a function with `{` but within that you open a block with `(`. Yeah ...))

Would be nice if the crate offers something like a transaction type, but otherwise we can create one.